Mathlib/Data/Matroid/Circuit.lean

```diff
@@ -166,4 +171,135 @@ lemma restrict_circuit_iff (hR : R ⊆ M.E := by aesop_mat) :
+    · rw [hI.mem_closure_iff]
+      exact .inl (hC.dep.superset hCss (insert_subset (hC.subset_ground heC) hI.subset_ground))
+    exact hC.not_indep (hI.subset (hCss.trans (by simp [heI])))
+  refine insert_subset heC (inter_subset_right.trans ?_)
+  refine (sInter_subset_of_mem (t := C \ {e}) ?_).trans diff_subset
+  simp [hCss, hC.mem_closure_diff_singleton_of_mem heC]
+
+/-! ### Dependence -/
+
+lemma Dep.exists_circuit_subset (hX : M.Dep X) : ∃ C, C ⊆ X ∧ M.Circuit C := by
+  obtain ⟨I, hI⟩ := M.exists_basis X
+  obtain ⟨e, heX, heI⟩ := exists_of_ssubset
+    (hI.subset.ssubset_of_ne (by rintro rfl; exact hI.indep.not_dep hX))
+  exact ⟨M.fundCircuit e I, (M.fundCircuit_subset_insert e I).trans (insert_subset heX hI.subset),
+    hI.indep.fundCircuit_circuit (hI.subset_closure heX) heI⟩
+
+lemma dep_iff_superset_circuit (hX : X ⊆ M.E := by aesop_mat) :
+    M.Dep X ↔ ∃ C, C ⊆ X ∧ M.Circuit C :=
+  ⟨Dep.exists_circuit_subset, fun ⟨C, hCX, hC⟩ ↦ hC.dep.superset hCX⟩
+
+lemma dep_iff_superset_circuit' : M.Dep X ↔ (∃ C, C ⊆ X ∧ M.Circuit C) ∧ X ⊆ M.E :=
+  ⟨fun h ↦ ⟨h.exists_circuit_subset, h.subset_ground⟩, fun ⟨⟨C, hCX, hC⟩, h⟩ ↦ hC.dep.superset hCX⟩
+
+lemma indep_iff_forall_subset_not_circuit' :
+    M.Indep I ↔ (∀ C, C ⊆ I → ¬M.Circuit C) ∧ I ⊆ M.E := by
+  simp_rw [indep_iff_not_dep, dep_iff_superset_circuit', not_and, imp_not_comm (b := _ ⊆ _)]; aesop
+
+lemma indep_iff_forall_subset_not_circuit (hI : I ⊆ M.E := by aesop_mat) :
+    M.Indep I ↔ ∀ C, C ⊆ I → ¬M.Circuit C := by
+  rw [indep_iff_forall_subset_not_circuit', and_iff_left hI]
+
+/-! ### Extensionality -/
+
+lemma ext_circuit {M₁ M₂ : Matroid α} (hE : M₁.E = M₂.E)
+    (h : ∀ ⦃C⦄, C ⊆ M₁.E → (M₁.Circuit C ↔ M₂.Circuit C)) : M₁ = M₂ := by
+  have h' : ∀ C, M₁.Circuit C ↔ M₂.Circuit C := by
+    exact fun C ↦ (em (C ⊆ M₁.E)).elim (h (C := C))
+      (fun hC ↦ iff_of_false (mt Circuit.subset_ground hC)
+        (mt Circuit.subset_ground (fun hss ↦ hC (hss.trans_eq hE.symm))))
+  refine ext_indep hE fun I hI ↦ ?_
+  simp_rw [indep_iff_forall_subset_not_circuit hI, h',
+    indep_iff_forall_subset_not_circuit (hI.trans_eq hE)]
+
+/-- A stronger version of `Matroid.ext_circuit`:
+two matroids on the same ground set are equal if no circuit of one is independent in another. -/
+lemma ext_circuit_not_indep {M₁ M₂ : Matroid α} (hE : M₁.E = M₂.E)
+    (h₁ : ∀ ⦃C⦄, M₁.Circuit C → ¬ M₂.Indep C) (h₂ : ∀ ⦃C⦄, M₂.Circuit C → ¬ M₁.Indep C) :
+    M₁ = M₂ := by
+  refine ext_circuit hE fun C _ ↦ ⟨fun hC ↦ ?_, fun hC ↦ ?_⟩
+  · have hC₂ : C ⊆ M₂.E := by rwa [← hE]
+    specialize h₁ hC
+    rw [not_indep_iff] at h₁
+    obtain ⟨C', hC'C, hC'⟩ := h₁.exists_circuit_subset
+    rwa [← hC.eq_of_not_indep_subset (h₂ hC') hC'C]
+  specialize h₂ hC
+  rw [not_indep_iff] at h₂
+  obtain ⟨C', hC'C, hC'⟩ := h₂.exists_circuit_subset
+  rwa [← hC.eq_of_not_indep_subset (h₁ hC') hC'C]
+
+lemma ext_iff_circuit {M₁ M₂ : Matroid α} :
+    M₁ = M₂ ↔ M₁.E = M₂.E ∧ ∀ ⦃C⦄, M₁.Circuit C ↔ M₂.Circuit C :=
```

Comment:
```suggestion
    M₁ = M₂ ↔ M₁.E = M₂.E ∧ ∀ C, M₁.Circuit C ↔ M₂.Circuit C :=
```
I think this would be easier to use, the key distinction between this situation and that in the previous two lemmas is that those have an explicit argument immediately following C which determines it, but here there is none such.